File: ruby/examples/error_handling.rb
# frozen_string_literal: true

# Branching on error type: local validation, authentication, and rate limits.
#
# Run:  PACTMAN_API_KEY=... bundle exec ruby examples/error_handling.rb

require "pactman/nonprofit_check_plus"

NCP = Pactman::NonprofitCheckPlus

api_key = ENV.fetch("PACTMAN_API_KEY", "")

if api_key.empty?
  warn "Set PACTMAN_API_KEY before running this example."
  exit 1
end

client = NCP::Client.new(api_key: api_key, base_url: ENV.fetch("PACTMAN_BASE_URL", nil), timeout: 10,
                         retry: { max_retries: 2 })

# One place to turn any SDK failure into an action. No string parsing.
def explain(error)
  case error
  when NCP::ValidationError
    detail = error.issues.empty? ? error.message : error.issues.map(&:message).join(" ")
    "Local validation — fix the input. #{detail}"
  when NCP::AuthenticationError
    "Authentication — the API key was rejected. Check PACTMAN_API_KEY."
  when NCP::RateLimitError
    "Rate limited — retry after #{error.retry_after_seconds || "an unspecified number of"} seconds."
  when NCP::TimeoutError
    "Timed out after #{error.timeout}s — raise the timeout or retry later."
  when NCP::NetworkError
    "Network failure — the request never reached the API."
  when NCP::ApiError
    "API error #{error.status} (request #{error.request_id || "unknown"}): #{error.message}"
  else
    "Unexpected: #{error}"
  end
end

# 1. A malformed EIN never leaves the process.
begin
  client.nonprofits.check("41178709")
rescue NCP::Error => e
  puts "malformed EIN -> #{explain(e)}"
end

# 2. An empty batch is rejected locally too.
begin
  client.nonprofits.check_bulk([])
rescue NCP::Error => e
  puts "empty batch   -> #{explain(e)}"
end

# 3. A bad key produces an authentication error on first use.
begin
  bad_client = NCP::Client.new(api_key: "obviously-not-a-real-key", base_url: ENV.fetch("PACTMAN_BASE_URL", nil),
                               retry: false)

  bad_client.nonprofits.check("411787097")
  puts "bad key       -> unexpectedly succeeded"
rescue NCP::Error => e
  puts "bad key       -> #{explain(e)}"
end

# 4. A real call, handled the same way.
begin
  result = client.nonprofits.check("41-1787097")
  puts "valid check   -> #{result.nonprofit&.organization_name || "no record"}"
rescue NCP::Error => e
  puts "valid check   -> #{explain(e)}"
end

File: java/examples/src/main/java/org/pactman/nonprofitcheckplus/examples/cases/ErrorHandling.java
package org.pactman.nonprofitcheckplus.examples.cases;

import java.util.Collections;
import java.util.stream.Collectors;
import org.pactman.nonprofitcheckplus.PactmanClient;
import org.pactman.nonprofitcheckplus.config.PactmanClientOptions;
import org.pactman.nonprofitcheckplus.config.RetryOptions;
import org.pactman.nonprofitcheckplus.examples.support.Example;
import org.pactman.nonprofitcheckplus.examples.support.ExampleContext;
import org.pactman.nonprofitcheckplus.examples.support.ExampleFailedException;
import org.pactman.nonprofitcheckplus.exceptions.PactmanApiException;
import org.pactman.nonprofitcheckplus.exceptions.PactmanAuthenticationException;
import org.pactman.nonprofitcheckplus.exceptions.PactmanNetworkException;
import org.pactman.nonprofitcheckplus.exceptions.PactmanRateLimitException;
import org.pactman.nonprofitcheckplus.exceptions.PactmanTimeoutException;
import org.pactman.nonprofitcheckplus.exceptions.PactmanValidationException;
import org.pactman.nonprofitcheckplus.exceptions.ValidationIssue;
import org.pactman.nonprofitcheckplus.models.SingleCheckResult;

/**
 * Branching on error type: local validation, authentication, and rate limits.
 */
public final class ErrorHandling implements Example {

    @Override
    public String id() {
        return "error-handling";
    }

    @Override
    public String title() {
        return "Branching on error type, with no string parsing";
    }

    @Override
    public void run(String[] args) {
        try (ExampleContext context = ExampleContext.withFixtures(
                PactmanClientOptions.builder()
                        .timeoutMs(10_000)
                        .retry(RetryOptions.builder().maxRetries(2).build()))) {
            PactmanClient client = context.client();

            // 1. A malformed EIN never leaves the process.
            try {
                client.nonprofits().check("41178709");
                System.out.println("malformed EIN -> unexpectedly succeeded");
            } catch (RuntimeException error) {
                System.out.println("malformed EIN -> " + explain(error));
            }

            // 2. An empty batch is rejected locally too.
            try {
                client.nonprofits().checkBulk(Collections.emptyList());
                System.out.println("empty batch   -> unexpectedly succeeded");
            } catch (RuntimeException error) {
                System.out.println("empty batch   -> " + explain(error));
            }

            // 3. A bad key produces an authentication error on first use.
            try (PactmanClient badClient = new PactmanClient(PactmanClientOptions.builder()
                    .apiKey("obviously-not-a-real-key")
                    .baseUrl(client.baseUrl())
                    .retry(RetryOptions.disabled())
                    .build())) {
                badClient.nonprofits().check("411787097");
                throw new ExampleFailedException("A rejected key should not have succeeded.");
            } catch (PactmanAuthenticationException error) {
                System.out.println("bad key       -> " + explain(error));
            }

            // 4. A real call, handled the same way.
            try {
                SingleCheckResult result = client.nonprofits().check("41-1787097");

                System.out.println("valid check   -> " + (result.getNonprofit() == null
                        ? "no record"
                        : result.getNonprofit().getOrganizationName()));
            } catch (RuntimeException error) {
                System.out.println("valid check   -> " + explain(error));
            }
        }
    }

    /** One place to turn any SDK failure into an action. No string parsing. */
    private static String explain(RuntimeException error) {
        if (error instanceof PactmanValidationException) {
            PactmanValidationException invalid = (PactmanValidationException) error;
            String detail = invalid.issues().isEmpty()
                    ? invalid.getMessage()
                    : invalid.issues().stream()
                            .map(ValidationIssue::message)
                            .collect(Collectors.joining(" "));

            return "Local validation — fix the input. " + detail;
        }

        if (error instanceof PactmanAuthenticationException) {
            return "Authentication — the API key was rejected. Check PACTMAN_API_KEY.";
        }

        if (error instanceof PactmanRateLimitException) {
            Double retryAfter = ((PactmanRateLimitException) error).retryAfterSeconds();

            return "Rate limited — retry after "
                    + (retryAfter == null ? "an unspecified" : retryAfter) + " seconds.";
        }

        if (error instanceof PactmanTimeoutException) {
            return "Timed out after " + ((PactmanTimeoutException) error).timeoutMs()
                    + "ms — raise timeoutMs or retry later.";
        }

        if (error instanceof PactmanNetworkException) {
            return "Network failure — the request never reached the API.";
        }

        if (error instanceof PactmanApiException) {
            PactmanApiException api = (PactmanApiException) error;

            return "API error " + api.status() + " (request "
                    + (api.requestId() == null ? "unknown" : api.requestId()) + "): "
                    + api.getMessage();
        }

        return "Unexpected: " + error;
    }
}
